lib/Storage/S3StorageClient.cpp

```diff
@@ -0,0 +1,379 @@
+
+  Aws::S3::Model::GetObjectRequest Request;
+  Request.SetBucket(Instance->Bucket);
+  Request.SetKey(IndexPath);
+
+  Aws::S3::Model::GetObjectOutcome Result = Instance->Client.GetObject(Request);
+  if (not Result.IsSuccess()) {
+    using Aws::Http::HttpResponseCode::NOT_FOUND;
+    if (Result.GetError().GetResponseCode() == NOT_FOUND)
+      return Instance;
+    else
+      return toError(Result);
+  }
+
+  std::string SerializedIndex;
+  {
+    constexpr std::size_t BufSize = 4096;
+    char Buffer[BufSize];
+    auto &Body = Result.GetResult().GetBody();
+
+    while (!Body.eof()) {
+      Body.read(Buffer, BufSize);
+      SerializedIndex.append(Buffer, Body.gcount());
+    }
+  }
+
+  llvm::yaml::Input YAMLInput(SerializedIndex);
+  YAMLInput >> Instance->FilenameMap;
+
+  return Instance;
+}
+
+std::string S3StorageClient::toString() const {
+  return SerializedString;
+}
+
+llvm::Expected<bool> S3StorageClient::exists(const llvm::StringRef Path) {
+  return FilenameMap.count(Path) != 0;
+}
+
+llvm::Error S3StorageClient::createDirectory(const llvm::StringRef Path) {
+  return llvm::Error::success();
+}
+
+llvm::Error S3StorageClient::remove(const llvm::StringRef Path) {
+  FilenameMap.erase(Path);
+  return llvm::Error::success();
+}
+
+llvm::sys::path::Style S3StorageClient::getStyle() const {
+  return llvm::sys::path::Style::posix;
+}
+
+llvm::Error S3StorageClient::copy(const llvm::StringRef Source,
+                                  const llvm::StringRef Destination) {
+  if (FilenameMap.count(Source) == 0)
+    return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                   "Source file %s does not exist",
+                                   Source.str().c_str());
+
+  FilenameMap[Destination] = FilenameMap[Source];
+  return llvm::Error::success();
+}
+
+llvm::Expected<ReadableFile>
+S3StorageClient::getReadableFile(const llvm::StringRef Path) {
+  using llvm::sys::fs::createTemporaryFile;
+  if (FilenameMap.count(Path) == 0)
+    return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                   "File %s does not exist",
+                                   Path.str().c_str());
+
+  Aws::S3::Model::GetObjectRequest Request;
+  Request.SetBucket(Bucket);
+  Request.SetKey(resolvePath(SubPath, FilenameMap[Path]));
+
+  Aws::S3::Model::GetObjectOutcome Result = Client.GetObject(Request);
+  if (not Result.IsSuccess())
+    return toError(Result);
+
+  llvm::SmallString<128> TempFile;
+  std::error_code EC = createTemporaryFile("revng-s3-storage-", "", TempFile);
+  if (EC)
+    return llvm::createStringError(EC, "Could not create temporary file");
+
+  std::ofstream OS(TempFile.str().str(), std::ios::binary);
+  if (OS.fail())
+    return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                   "Could not open temporary file");
+
+  constexpr std::size_t BufSize = 4096;
+  char Buffer[BufSize];
```

Comment:
Don't put things so big on the stack. Use a `std::vector`.

Reply:
Switched to `llvm::SmallVector` since it has `resize_for_overwrite`